### PR TITLE
Switch from SQLTypeInfo to hdk::ir::Type (part 5)

### PR DESCRIPTION
**Pull request by ienkovich**
Add hdk::ir::Type fields to TargetMetaInfo and TargetInfo.

### CODE REVIEW

File: IR/TypeUtils.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+#include "Type.h"
+
+namespace hdk::ir {
+
+/**
+ * Return a logical type for the given type. That assumes following
+ * modifications to the type (if applicable):
+ *  - size for datetime types is set to 8 bytes
+ *  - date type time unit is set to kSecond
+ *  - external dictionary type size is set to 4 bytes
```
Comment (kurapov-peter) [inline IR/TypeUtils.h:18]: What should happen to an external dictionary of size 8? The code seem to just return the original type. Maybe reformulate? Sounds a bit misleading.

**Review by kurapov-peter — COMMENTED** [review]
Just a comment and a general note: having logical/physical and just types represented by a single `Type` seem a bit confusing. What's the reasoning behind not having them separated? Not to have a lot of duplicated code?

**Comment by ienkovich** [comment]
> Just a comment and a general note: having logical/physical and just types represented by a single `Type` seem a bit confusing. What's the reasoning behind not having them separated? Not to have a lot of duplicated code?

Why are you confused about using the same structure to store logical and physical types? In most cases, they are equal.

File: IR/TypeUtils.h (hunk 1 above)
Comment (ienkovich) [reply IR/TypeUtils.h:18]: External dictionaries always have size <= 4.

File: IR/TypeUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "TypeUtils.h"
+
+#include "Context.h"
+
+namespace hdk::ir {
+
+const Type* logicalType(const Type* type) {
```
Comment (alexbaden) [inline IR/TypeUtils.cpp:13]: I find this method less readable than previous method. 

Essentially logical type strips away the encoding and returns the type as if it had no encoding. This is useful for determining the output slot size (output types are not encoded) or for other places where copying of the data in a type may occur. 

So, two questions -- are we doing things other than stripping encoding here? And, if not, perhaps we can rename this or add it to the context object (perhaps something like `Context::stripEncoding(Type *)`). Are there LLVM analogues for what we're doing here?

File: QueryEngine/TargetMetaInfo.h [hunk 3]
```diff
@@ -19,21 +19,32 @@
...
+      : resname_(resname)
+      , type_(type)
+      , physical_type_(physical_type)
+      , ti_(type_->toTypeInfo())
+      , physical_ti_(physical_type_->toTypeInfo()) {}
   const std::string& get_resname() const { return resname_; }
+  const hdk::ir::Type* type() const { return type_; }
+  const hdk::ir::Type* physicalType() const { return physical_type_; }
   const SQLTypeInfo& get_type_info() const { return ti_; }
   const SQLTypeInfo& get_physical_type_info() const { return physical_ti_; }
```
Comment (alexbaden) [inline QueryEngine/TargetMetaInfo.h:49]: This appears to be un-used. I think it is left over from geo? Perhaps we can just remove it now (and the newly added `physicalType()` above)?

**Review by alexbaden — CHANGES_REQUESTED** [review]

File: IR/TypeUtils.cpp (hunk 2 above)
Comment (ienkovich) [reply IR/TypeUtils.cpp:13]: I don't think there are analogs in LLVM or Arrow for that. There is just no such thing as encoding attached to a type. And in OmniSci encoding means multiple things, it is time unit, value size, and dictionary usage. Also, the original `get_logical_type` resets size, which affects fixed-length arrays.

In OmniSci date in days is considered to be an encoded version of date in seconds. In Arrow and our new type system they are just two date types with different time units. Why should I get date in seconds when selecting date in days? I guess it's just to make the engine simpler. This types mapping is very OmniSci specific and depends on what types are considered to be basic and are supported in output. 

I'd like to avoid `encoding` words here because we don't introduce any encoding terms in our type system. I'm not a fan of the current `logicalType` naming either, just used the existing one. Probably, `targetType` or `outputType` or `resultType`?

I didn't put this method into `Type` or `Context` class because I'm not sure it has to be public. It shouldn't be required for users to build expressions.

File: QueryEngine/TargetMetaInfo.h (hunk 3 above)
Comment (ienkovich) [reply QueryEngine/TargetMetaInfo.h:49]: Yes, you are right, it can be removed.

File: IR/TypeUtils.cpp (hunk 2 above)
Comment (alexbaden) [reply IR/TypeUtils.cpp:13]: The date in days vs date in seconds is broken, they should indeed be separate types. It was an implementation issue that got pushed out and would have been very difficult to mitigate on the on-disk data that was already created. 

What do you mean by not introducing encoding terms in our type system? Don't we have a 32-bit, 16-bit, and 8-bit dictionary encoded string? Those are all encodings for the TEXT type.

Comment (alexbaden) [reply IR/TypeUtils.cpp:13]: I see, those are all different types. 

`commonType`? Also not great...

Comment (ienkovich) [reply IR/TypeUtils.cpp:13]: I mean the user can easily misunderstand what the word `encoding` means in the context of our type system.

We can say that DICT32 type is encoding of TEXT type, but somehow `stripEncoding` wouldn't return TEXT, it would still return DICT32. At the same time, it would return TIME64 for TIME32, but wouldn't return INT64 for INT32. For dates time unit is affected by this method, for timestamps - not.

Basically, this method returns a type that is internally used for processing the data of the source type. So, might be better to build a naming out of that.

Comment (ienkovich) [reply IR/TypeUtils.cpp:13]: Probably, `canonicalType`? Since it's common to transform data to a canonical form prior to further processing.

Comment (alexbaden) [reply IR/TypeUtils.cpp:13]: I suppose that's fine. 

What if we made it a no-op attached to the Type? E.g. you would call `Type::canonicalize()` which would return either the exact same type, or a new type if needed (e.g. in one of the cases above). We would then replace all locations where we call `get_logical_type_info` with `canonicalize()`. There are only about 12 of them. That way, if you add a new type, you are prodded to properly define canonicalize, and we eliminate the branching when we likely already know the type (although in reality that probably doesn't add any noticeable overhead).

Comment (ienkovich) [reply IR/TypeUtils.cpp:13]: Later patches also add `logicalSize`. Which is not equal to `logicalType(...)->size()` for arrays. I'll check if these two actually may be synchronized. Size reset for fixed-length arrays is probably a non-desired effect in `get_logical_type_info`.

Comment (ienkovich) [reply IR/TypeUtils.cpp:13]: To avoid rebase of all existing branches, I added this change as a separate part. It replaces `logicalType` and `logicalSize` methods with `Expr::canonicalize` and `Expr::canonicalSize`. It will be part 45.
